src/CLIFramework/Extension/DaemonExtension.php

```diff
@@ -0,0 +1,203 @@
+    }
+
+    /**
+     * Call this method if you want to change the current directory on making a daemon process.
+     */
+    protected function changeDirectory()
+    {
+        $this->chdir = true;
+    }
+
+    public function getPidFilePath()
+    {
+        if ($this->getCommandOptions() && $this->getCommandOptions()->{'pid-file'}) {
+            return $this->getCommandOptions()->{'pid-file'};
+        }
+        $pid = getmypid();
+        $pidFile = $this->command ? $this->command->getName() : $pid;
+        return $this->config->getPidDirectory() . "/$pidFile.pid";
+    }
+
+    private function bindHooks($command)
+    {
+        $extension = $this;
+        $command->addHook('execute.before', function() use ($extension) {
+            $extension->run();
+        });
+        $command->addHook('execute.after', function() use ($extension) {
+            @unlink($extension->getPidFilePath());
+        });
+    }
+
+    private function bindOptions($command)
+    {
+        $options = $command->getOptionCollection();
+        $options->add('pid-file?', 'The path of pid file.');
+    }
+
+    private function prepareLogger()
+    {
+        $logPath = $this->getLogPath();
+        $logger = $this->getLogger();
+
+        if (!$logPath || !$logger) {
+            return;
+        }
+
+        $resource = fopen($logPath, "a+");
+
+        if ($resource === false) {
+            throw new ExtensionException("Can't open file: $logPath");
+        }
+
+        // TODO change logging style
+        $logger->setWriter(new StreamWriter($resource));
+    }
+
+    private function daemonize()
+    {
+        switch (pcntl_fork()) {
+        case -1:
+            throw new ExtensionException("pcntl_fork() failed");
+        case 0:
+            break;
+        default:
+            exit(0);
+        }
+
+        if (!$this->createPidFile()) {
+            throw new ExtensionException("creating a pid file failed");
+        }
+
+        if ($this->chdir) {
+            $this->chdir();
+        }
+
+        if ($this->detach) {
+            $this->detach();
+        }
+
+    }
+
+    private function chdir()
+    {
+        if ($this->chdir && !chdir("/")) {
+            throw new ExtensionException("chdir failed");
+        }
+    }
+
+    private function detach()
+    {
+        if (!fclose(STDIN)) {
+            throw new ExtensionException("fclose(STDIN) failed");
+        }
+
+        if (!fclose(STDOUT)) {
+            throw new ExtensionException("fclose(STDOUT) failed");
+        }
+
+        if (!fclose(STDERR)) {
+            throw new ExtensionException("fclose(STDERR) failed");
+        }
+    }
+
+    private function createPidFile()
+    {
+        return file_put_contents($this->getPidFilePath(), getmypid());
+    }
+
+    private function getLogPath()
+    {
+        $options = $this->getApplicationOptions();
+        return $options && $options->{'log-path'} ? $options->{'log-path'} : null;
+    }
+
+    private function getLogger()
```

Comment:
`$this->logger` can be retrieved from service container. we should move this into the constructor of this class. because `command->getLogger()` is basically the same as `$container['logger']`. we should make them consistent.